src/cli.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+#![allow(unused_variables)]
+#![allow(dead_code)]
+
+use crate::errors::DenoResult;
+use crate::isolate_init::IsolateInit;
+use crate::isolate_state::IsolateState;
+use crate::msg_ring;
+use crate::ops;
+use crate::permissions::DenoPermissions;
+use deno_core::deno_buf;
+use deno_core::deno_mod;
+use deno_core::Behavior;
+use deno_core::Op;
+use std::cell::Cell;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::time::Instant;
+
+// Buf represents a byte array returned from a "Op". The message might be empty
+// (which will be translated into a null object on the javascript side) or it is
+// a heap allocated opaque sequence of bytes.  Usually a flatbuffer message.
+pub type Buf = Box<[u8]>;
+
+pub type CliOp = Op<Buf>;
+
+/// Implements deno_core::Behavior for the main Deno command-line.
+pub struct Cli {
+  shared: Option<deno_buf>, // Pin?
+  tx: msg_ring::Sender,
+  rx: msg_ring::Receiver,
+  init: IsolateInit,
+  timeout_due: Cell<Option<Instant>>,
+  pub state: Arc<IsolateState>,
+  pub permissions: Arc<DenoPermissions>, // TODO(ry) move to IsolateState
+}
+
+impl Cli {
+  pub fn new(
+    init: IsolateInit,
+    state: Arc<IsolateState>,
+    permissions: DenoPermissions,
+  ) -> Self {
+    let buffer = msg_ring::Buffer::new(1024 * 1024);
```

Comment:
This buffer is too small for some tests to complete. Might need a redesign.
Changing it to `1024 * 8196` bytes(~8MB) get us past the current failure, but errors out on a later test with:
```
tests/error_004_missing_module.test ... FAILED
Expected output does not match actual.
Expected output: 
Compiling [WILDCARD]tests/error_004_missing_module.ts
Uncaught NotFound: Cannot resolve module "bad-module.ts" from "[WILDCARD]/tests/error_004_missing_module.ts"
    at DenoError ([WILDCARD]/js/errors.ts:[WILDCARD])
    at maybeError ([WILDCARD]/js/errors.ts:[WILDCARD])
    at maybeThrowError ([WILDCARD]/js/errors.ts:[WILDCARD])
    at sendSync ([WILDCARD]/js/dispatch.ts:[WILDCARD])
    at fetchModuleMetaData ([WILDCARD]/js/os.ts:[WILDCARD])
    at _resolveModule ([WILDCARD]/js/compiler.ts:[WILDCARD])
    at resolveModuleNames ([WILDCARD]/js/compiler.ts:[WILDCARD])
    at compilerHost.resolveModuleNames ([WILDCARD]typescript.js:[WILDCARD])
    at resolveModuleNamesWorker ([WILDCARD]typescript.js:[WILDCARD])
    at resolveModuleNamesReusingOldState ([WILDCARD]typescript.js:[WILDCARD])

Actual output:   
Compiling file:///home/andy/Documents/projects/personal/deno/tests/error_004_missing_module.ts
Uncaught NotFound: Cannot resolve module "bad-module.ts" from "/home/andy/Documents/projects/personal/deno/tests/error_004_missing_module.ts"
    at DenoError (deno/js/errors.ts:22:5)
    at maybeError (deno/js/errors.ts:33:12)
    at maybeThrowError (deno/js/errors.ts:39:15)
    at sendSync (deno/js/dispatch.ts:87:5)
    at fetchModuleMetaData (deno/js/os.ts:81:19)
    at _resolveModule (deno/js/compiler.ts:249:38)
    at resolveModuleNames (deno/js/compiler.ts:479:35)
    at compilerHost.resolveModuleNames (deno/third_party/node_modules/typescript/lib/typescript.js:118650:138)
    at resolveModuleNamesWorker (deno/third_party/node_modules/typescript/lib/typescript.js:86767:127)
    at resolveModuleNamesReusingOldState (deno/third_party/node_modules/typescript/lib/typescript.js:87001:24)thread '
tokio-runtime-worker-0' panicked at 'called `Result::unwrap()` on an `Err` value: DenoError { repr: Simple(BadResource, "bad resource id") }', src/libcore/result.rs:997:5
```

Reply:
Ah - thanks! I want to get the error messages when this happens to be more useful...

Reply:
It looks like this might become problematic. The record that failed in that test was smaller than 1048576 bytes, so it was likely caused by a chain of blocking operations. At what point are records removed from the buffer here(I assume that is what `records_shift` is for)? Could this garbage collection potentially be blocked by chain operations(I.E. lots of imports in a single file)?